feat: Add streaming API to orchestration (#352)
SAP/ai-sdk-js: packages/orchestration/README.md

}
```

Each chunk will be a string containing the delta content.

#### Streaming with Abort Controller

Streaming request can be aborted using the `AbortController` API.
In case of an error, the SAP Cloud SDK for AI will automatically close the stream.
Additionally, it can be aborted manually by calling the `stream()` method with an `AbortController` object.

```ts
const orchestrationClient = new OrchestrationClient({
  llm: {
    model_name: 'gpt-4o',
    model_params: { max_tokens: 50, temperature: 0.1 }
  },
  templating: {
    template: [
      { role: 'user', content: 'Give a long history of {{?country}}?' }
    ]
  }
});

const controller = new AbortController();
const response = await orchestrationClient.stream(
  {
    inputParams: { country: 'France' }
  },
  controller
);

// Abort the streaming request after one second
setTimeout(() => {
  controller.abort();
}, 1000);

for await (const chunk of response.stream) {
  console.log(JSON.stringify(chunk));
}
```

In this example, streaming request will be aborted after one second.
Abort controller can be useful, e.g., when end-user wants to stop the stream or refreshes the page.

#### Stream Options

> The orchestration service offers multiple streaming options, which you can configure in addition to the LLM's streaming options.

Reviewer: [pp] multiple streaming options such as?

> There are two ways to add specific streaming options to your client, either at initalization, or dynamically when calling the stream API.

Reviewer: ```suggestion
There are two ways to add specific streaming options to your client, either at initialization of orchestration client, or when calling the stream API.
```

> Dynamically setting these options after client initialization is particularly helpful when you've initialized a client with a config meant for regular chat completion and now want to switch to using streaming.

Reviewer: ```suggestion
Setting streaming options dynamically could be useful if an initialized orchestration client will also be used for streaming.
```

But I mean even if a user put stream options when initializing the client, they can still use the client for non-streaming chat completion right? I can only imagine setting stream options dynamically if user write some kind of playground or reuse the same orchestration client for different use cases (which requires e.g. different filter settings?)

You can check the list of available stream options in the [orchestration service's documentation](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/streaming).

An example for setting the streaming options when calling the stream API looks like the following:

```ts
const response = orchestrationClient.stream(
  {
    inputParams: { country: 'France' }
  },
  controller,
  {
    llm: { include_usage: false },
    global: { chunk_size: 10 },
    outputFiltering: { overlap: 200 }
  }
);
```

> Usage metrics are collected by default, if you do not want to receive them, set include_usage to false.

Reviewer: ```suggestion
Usage metrics are returned by default, if you do not want to receive them, set `include_usage` to `false`.
```

> If you don't want any streaming options as part of your call to the LLM, set options.llm = null.

Reviewer: ```suggestion
If you don't want any streaming options as part of your call to the LLM, set `streamOptions.llm` to `null`.
```

}
```

In this example, streaming request will be aborted after one second.
Abort controller can be useful, e.g., when end-user wants to stop the stream or refreshes the page.

#### Stream Options

The orchestration service offers multiple streaming options, which you can configure in addition to the LLM's streaming options.
These include options like definining the maximum number of characters per chunk or modifying the output filter behavior.
There are two ways to add specific streaming options to your client, either at initialization of orchestration client, or when calling the stream API.

Setting streaming options dynamically could be useful if an initialized orchestration client will also be used for streaming.

You can check the list of available stream options in the [orchestration service's documentation](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/streaming).

An example for setting the streaming options when calling the stream API looks like the following:

```ts
const response = orchestrationClient.stream(
  {
    inputParams: { country: 'France' }
  },
  controller,
  {
    llm: { include_usage: false },
    global: { chunk_size: 10 },

>     outputFiltering: { overlap: 200 }

Reviewer: [q] Out of curiosity, have you tried setting streaming options for filtering without providing a filter?

Author: I did 👍
